assets/app/scripts/services/data.js

```diff
@@ -360,6 +360,122 @@ angular.module('openshiftConsole')
     return deferred.promise;
   };
 
+// https://developer.mozilla.org/en-US/docs/Web/API/WindowBase64/btoa
+function utf8_to_b64( str ) {
+    return window.btoa(window.unescape(encodeURIComponent( str )));
+}
+function b64_to_utf8( str ) {
+    return decodeURIComponent(window.escape(window.atob( str )));
+}
+
+// TODO (bpeterse): ideally DataService would be more general & log specific function
+// could be in a separate service, but we need the helper functions
+DataService.prototype.createStream = function(kind, name, context, isBinary) {
+  var getNamespace = this._getNamespace.bind(this);
+  var urlForResource = this._urlForResource.bind(this);
+  kind = this.kindToResource(kind) ?
+              this.kindToResource(kind) :
+              normalizeResource(kind);
+
+  var protocols = isBinary ? 'binary.k8s.io' : 'base64.binary.k8s.io';
+  var identifier = 'stream_';
+  var openQueue = {};
+  var messageQueue = {};
+  var closeQueue = {};
+  var errorQueue = {};
+
```

Comment:
Bringing the convo about changing the queues to use `$.Callbacks()` here (out of irc).  So:

```javascript
// current
var handler1 = stream.open(function() {});
var handler2 = stream.message(function() {});
var handler3 = stream.close(function() {});
stream.remove(handler1);
stream.remove(handler2);
stream.remove(handler3);
```
Since `$.Callbacks()` uses the fn itself as the handler, rather than a unique id, my public API would have to change.  Exploring a few ideas:

```javascript
// a dummy fn for examples
var fn = function() {
  // do stuff.
}

// 1.)
// the current simple remove() no longer works with $.Callbacks()
streamer.onOpen(fn);
streamer.onMessage(fn);
// ...remove from open? or message? or both?
streamer.remove(fn);


// 2.)
// this would work better... but i dislike using strings :(
streamer.on('open', fn);
streamer.on('message', fn);
streamer.remove('message', fn);


// 3.)
// im good with functions returning functions, but var unRegister looks weird & it feels a little wonky.
var unRegister = streamer.message(fn);
unRegister();

// 4.) 
// returning an object w/a method bound to the fns registry is ok, 
// but is not symmetrical
var handler = streamer.message(fn);
handler.deRegister();
```
I'm not completely sold on any option. I'd probably prefer to (eventually) refactor `jQuery` out of `DataService` & only use it in Directives, where we absolutely need additional DOM manipulation beyond what `angular.element` provides & not use `$.Callbacks` here.  I mentioned an example of a [native angular pubsub](https://github.com/benjaminapetersen/angular-simple-pubsub) module in irc as a possibility to explore (later).

@jwforres thoughts?